### DISCUSSION TITLE
PyObject_CallOneArg?

### DISCUSSION THREAD

**Discussion by tjstum** [wjakob/nanobind#118] [Q&A] [answered]
Hi there!

While studying the [implementation](https://github.com/wjakob/nanobind/blob/947fbb071eca282e6350402c0a23800502a1644f/src/nb_type.cpp#L682) of implicit conversions (in `nb_type.cpp`), I came across the new (to me!) vector call:
```c++
    PyObject *args[2] = { nullptr, src };
    result = PyObject_Vectorcall((PyObject *) dst_type->type_py, args + 1,
                                 NB_VECTORCALL_ARGUMENTS_OFFSET + 1, nullptr);
```
When I was reading the [docs](https://docs.python.org/3/c-api/call.html), I also noticed that there is a convenience method [`PyObject_CallOneArg`](https://docs.python.org/3/c-api/call.html#c.PyObject_CallOneArg) that [appears](https://github.com/python/cpython/blob/28db978d7f134edf6c86f21c42e15003511e7e9b/Objects/call.c#L370) to do the same thing.

So I'm left with this feeling that the nanobind call in nb_type is intentionally using `PyObject_VectorCall` instead of `PyObject_CallOneArg` but don't see why.
I did some more digging, and I found a commit (https://github.com/wjakob/nanobind/commit/a6739ba2e4290d704766baab1ddcbe1751381812) that changed from `PyObject_CallOneArg` to `NB_VECTORCALL`. That makes sense since 3.8 didn't have `PyObject_CallOneArg`. But Python also didn't have `PyObject_VectorCall` until 3.9 (looks like NB_VECTORCALL would also use the provisional version). And now, the `nb_type.cpp` code above is in a snippet that only runs on 3.9+ (after the change to support the limited API).

So I'm wondering if there's any current reason to use `PyObject_VectorCall` and construct the array directly, instead of using `PyObject_CallOneArg`. I'm certainly not advocating for one or the other! I'm just looking to broaden my understanding.

Thanks!

**Comment by wjakob** [comment, marked as answer]
Hi @tjstum. The code that you see right now is the result of quite a bit of evolution, and initially `PyObject_CallOneArg` was not used because it's unavailable on Python 3.8 (before the ifdef you mentioned was added). Looking at the implementation of `PyObject_CallOneArg` in CPython, I don't think there is a performance advantage over just directly using `PyObject_VectorCall`. This explains why it was not changed at some point.

**Reply by tjstum** [reply]
Cool, thanks for the explanation!
